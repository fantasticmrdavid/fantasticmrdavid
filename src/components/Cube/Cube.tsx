import React, { memo } from "react";
import { BorderTracer } from "components/BorderTracer/BorderTracer";
import {
  Container,
  Polygon,
  Front,
  Back,
  Left,
  Right,
  Top,
  Bottom,
} from "./cube.styles";

export interface CubeProps {
  isLoading: boolean;
}

export const Cube = memo(({ isLoading }: CubeProps) => (
  <Container $isLoading={isLoading}>
    <Polygon>
      <Front>
        <BorderTracer direction="left" cycleColor />
      </Front>
      <Back>
        <BorderTracer cycleColor />
      </Back>
      <Top />
      <Bottom>
        <BorderTracer cycleColor />
      </Bottom>
      <Left>
        <BorderTracer direction="left" cycleColor />
      </Left>
      <Right>
        <BorderTracer cycleColor />
      </Right>
    </Polygon>
  </Container>
));
