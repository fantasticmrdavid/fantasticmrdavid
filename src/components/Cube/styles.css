.Cube_container {
  position: absolute;
  bottom: 0px;
  right: 0px;
	perspective: 800px;
	perspective-origin: 50% 150px;
  transition: 0.5s;
  transform: scale(2);

  &.loading {
    transform: scale(8);
  }
}

.Cube {
	position: absolute;;
  bottom: 225px;
  right: -100px;
	width: 300px;
	transform-style: preserve-3d;
  animation: spin 0.6s normal linear;
  animation-fill-mode: forwards;

  div {
  	position: absolute;
  	width: 300px;
  	height: 300px;
    background: rgba(209, 43, 43, 0.1);
    border: 1px solid rgba(209, 43, 43, 0.2);
  }

  .back {
  	transform: translateZ(-150px) rotateY(180deg);
  }
  .right {
  	transform: rotateY(-270deg) translateX(150px);
  	transform-origin: top right;
  }
  .left {
  	transform: rotateY(270deg) translateX(-150px);
  	transform-origin: center left;
  }
  .top {
  	transform: rotateX(-90deg) translateY(-150px);
  	transform-origin: top center;
  }
  .bottom {
  	transform: rotateX(90deg) translateY(150px);
  	transform-origin: bottom center;
  }
  .front {
  	transform: translateZ(150px);
  }
}

@keyframes spin {
	from { transform: rotateX(15deg) rotateY(15deg); }
	to { transform: rotateX(-60deg) rotateY(60deg); }
}
