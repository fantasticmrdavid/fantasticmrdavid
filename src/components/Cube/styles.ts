import styled, { css, keyframes } from 'styled-components';
import { transparentize } from 'polished';

const cubeColors = [
  'rgb(209, 43, 43)',
  'rgb(66, 33, 99)',
  'rgb(0, 0, 255)',
];

interface ContainerProps {
  loading: boolean,
}

const spin = keyframes`
  from { transform: rotateX(15deg) rotateY(15deg); }
  to { transform: rotateX(-60deg) rotateY(60deg); }
`;

const cycleColor = keyframes`
  ${cubeColors.map((c: string, i: number) => `
    ${i * (100 / cubeColors.length)}% {
      background-color: ${transparentize(0.9, c)};
      border: ${`1px solid ${transparentize(0.8, c)}`};
    }`)}
    100% {
      background-color: ${transparentize(0.9, cubeColors[0])};
      border: ${`1px solid ${transparentize(0.8, cubeColors[0])}`};
    }
`;

export const Container = styled.div`
  position: absolute;
  bottom: 0px;
  right: 0px;
  perspective: 800px;
  perspective-origin: 50% 150px;
  transition: 0.5s;
  transform: ${css`scale(${(props: ContainerProps) => (props.loading ? 8 : 2)})`};
`;

export const Polygon = styled.div`
  position: absolute;
  bottom: 225px;
  right: -100px;
  width: 300px;
  transform-style: preserve-3d;
  animation: ${css`${spin} 0.6s normal linear`};
  animation-fill-mode: forwards;
`;

const Side = styled.div`
  position: absolute;
  width: 300px;
  height: 300px;
  animation: ${css`${cycleColor} 45s normal infinite`};
`;

export const Back = styled(Side)`
  transform: translateZ(-150px) rotateY(180deg);
`;

export const Right = styled(Side)`
  transform: rotateY(-270deg) translateX(150px);
  transform-origin: top right;
`;

export const Left = styled(Side)`
  transform: rotateY(270deg) translateX(-150px);
  transform-origin: center left;
`;

export const Top = styled(Side)`
  transform: rotateX(-90deg) translateY(-152px);
  transform-origin: top center;
`;

export const Bottom = styled(Side)`
  transform: rotateX(90deg) translateY(150px);
  transform-origin: bottom center;
`;

export const Front = styled(Side)`
  transform: translateZ(150px);
`;
