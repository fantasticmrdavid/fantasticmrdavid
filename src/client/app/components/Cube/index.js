import React, { PropTypes } from 'react';
import styles from './styles.css';

const Cube = (props) => {
  const { loading } = props;
  return (
    <div className={`${styles.Cube_container} ${!!loading && styles.loading}`}>
      <div className={styles.Cube}>
        <div className={styles.front} />
        <div className={styles.back} />
        <div className={styles.top} />
        <div className={styles.bottom} />
        <div className={styles.left} />
        <div className={styles.right} />
      </div>
    </div>
  );
};

Cube.propTypes = {
  loading: PropTypes.bool,
};

export default Cube;
