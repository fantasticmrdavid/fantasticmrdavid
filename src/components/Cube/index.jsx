import React from 'react';
import PropTypes from 'prop-types';
import * as styles from './styles';

const Cube = ({ loading }) => {
  const {
    Container,
    Polygon,
    Front,
    Back,
    Left,
    Right,
    Top,
    Bottom,
  } = styles;

  return (
    <Container loading={loading}>
      <Polygon>
        <Front />
        <Back />
        <Top />
        <Bottom />
        <Left />
        <Right />
      </Polygon>
    </Container>
  );
};

Cube.propTypes = {
  loading: PropTypes.bool,
};

export default Cube;
