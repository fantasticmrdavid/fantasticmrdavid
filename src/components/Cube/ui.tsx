import React from 'react';
import * as styles from './styles';

interface Props {
  loading: boolean,
}

export default ({ loading }: Props) => {
  const {
    Container,
    Polygon,
    Front,
    Back,
    Left,
    Right,
    Top,
    Bottom,
  } = styles;

  return (
    <Container loading={loading}>
      <Polygon>
        <Front />
        <Back />
        <Top />
        <Bottom />
        <Left />
        <Right />
      </Polygon>
    </Container>
  );
};
