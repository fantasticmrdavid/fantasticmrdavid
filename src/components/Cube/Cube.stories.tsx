import React from "react";
import { Cube } from "./Cube";
import { ComponentStory, Meta } from "@storybook/react";

export default {
  title: "Cube",
  component: Cube
} as Meta

const Template: ComponentStory<typeof Cube> = (args) => <div style={{ width: "100%", height: "100%", position: "relative" }}><Cube {...args} /></div>;

export const Loading = Template.bind({ isLoading: true });
export const Loaded = Template.bind({ isLoading: false });
