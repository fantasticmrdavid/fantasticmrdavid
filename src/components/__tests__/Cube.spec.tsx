import React from 'react';
import { render } from "@testing-library/react";
import Cube from 'components/Cube';

describe('Component -> Cube', () => {
  it('should render the Cube in loading state without throwing an error', () => {
    expect(render(<Cube isLoading />)).toMatchSnapshot();
  });

  it('should render the Cube in completed state without throwing an error', () => {
    expect(render(<Cube isLoading={false} />)).toMatchSnapshot();
  });
});
