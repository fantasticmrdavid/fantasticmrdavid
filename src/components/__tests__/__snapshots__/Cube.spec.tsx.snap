// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component -> Cube should render the Cube in completed state without throwing an error 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="styles__Container-sc-1pwogyu-0 fhwLOl"
      >
        <div
          class="styles__Polygon-sc-1pwogyu-1 bFqjCw"
        >
          <div
            class="styles__Side-sc-1pwogyu-2 styles__Front-sc-1pwogyu-8 fiFRhu bmLGPK"
          >
            <div
              class="styles__Tracer-sc-18vfsm6-0 geFGiM"
              direction="left"
            />
          </div>
          <div
            class="styles__Side-sc-1pwogyu-2 styles__Back-sc-1pwogyu-3 fiFRhu hPnqZU"
          >
            <div
              class="styles__Tracer-sc-18vfsm6-0 eDtdvo"
            />
          </div>
          <div
            class="styles__Side-sc-1pwogyu-2 styles__Top-sc-1pwogyu-6 fiFRhu vbVpk"
          />
          <div
            class="styles__Side-sc-1pwogyu-2 styles__Bottom-sc-1pwogyu-7 fiFRhu cjImYD"
          >
            <div
              class="styles__Tracer-sc-18vfsm6-0 eDtdvo"
            />
          </div>
          <div
            class="styles__Side-sc-1pwogyu-2 styles__Left-sc-1pwogyu-5 fiFRhu enKYWw"
          >
            <div
              class="styles__Tracer-sc-18vfsm6-0 geFGiM"
              direction="left"
            />
          </div>
          <div
            class="styles__Side-sc-1pwogyu-2 styles__Right-sc-1pwogyu-4 fiFRhu kZyHbZ"
          >
            <div
              class="styles__Tracer-sc-18vfsm6-0 eDtdvo"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="styles__Container-sc-1pwogyu-0 fhwLOl"
    >
      <div
        class="styles__Polygon-sc-1pwogyu-1 bFqjCw"
      >
        <div
          class="styles__Side-sc-1pwogyu-2 styles__Front-sc-1pwogyu-8 fiFRhu bmLGPK"
        >
          <div
            class="styles__Tracer-sc-18vfsm6-0 geFGiM"
            direction="left"
          />
        </div>
        <div
          class="styles__Side-sc-1pwogyu-2 styles__Back-sc-1pwogyu-3 fiFRhu hPnqZU"
        >
          <div
            class="styles__Tracer-sc-18vfsm6-0 eDtdvo"
          />
        </div>
        <div
          class="styles__Side-sc-1pwogyu-2 styles__Top-sc-1pwogyu-6 fiFRhu vbVpk"
        />
        <div
          class="styles__Side-sc-1pwogyu-2 styles__Bottom-sc-1pwogyu-7 fiFRhu cjImYD"
        >
          <div
            class="styles__Tracer-sc-18vfsm6-0 eDtdvo"
          />
        </div>
        <div
          class="styles__Side-sc-1pwogyu-2 styles__Left-sc-1pwogyu-5 fiFRhu enKYWw"
        >
          <div
            class="styles__Tracer-sc-18vfsm6-0 geFGiM"
            direction="left"
          />
        </div>
        <div
          class="styles__Side-sc-1pwogyu-2 styles__Right-sc-1pwogyu-4 fiFRhu kZyHbZ"
        >
          <div
            class="styles__Tracer-sc-18vfsm6-0 eDtdvo"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Component -> Cube should render the Cube in loading state without throwing an error 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="styles__Container-sc-1pwogyu-0 fhwLOl"
      >
        <div
          class="styles__Polygon-sc-1pwogyu-1 bFqjCw"
        >
          <div
            class="styles__Side-sc-1pwogyu-2 styles__Front-sc-1pwogyu-8 fiFRhu bmLGPK"
          >
            <div
              class="styles__Tracer-sc-18vfsm6-0 geFGiM"
              direction="left"
            />
          </div>
          <div
            class="styles__Side-sc-1pwogyu-2 styles__Back-sc-1pwogyu-3 fiFRhu hPnqZU"
          >
            <div
              class="styles__Tracer-sc-18vfsm6-0 eDtdvo"
            />
          </div>
          <div
            class="styles__Side-sc-1pwogyu-2 styles__Top-sc-1pwogyu-6 fiFRhu vbVpk"
          />
          <div
            class="styles__Side-sc-1pwogyu-2 styles__Bottom-sc-1pwogyu-7 fiFRhu cjImYD"
          >
            <div
              class="styles__Tracer-sc-18vfsm6-0 eDtdvo"
            />
          </div>
          <div
            class="styles__Side-sc-1pwogyu-2 styles__Left-sc-1pwogyu-5 fiFRhu enKYWw"
          >
            <div
              class="styles__Tracer-sc-18vfsm6-0 geFGiM"
              direction="left"
            />
          </div>
          <div
            class="styles__Side-sc-1pwogyu-2 styles__Right-sc-1pwogyu-4 fiFRhu kZyHbZ"
          >
            <div
              class="styles__Tracer-sc-18vfsm6-0 eDtdvo"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="styles__Container-sc-1pwogyu-0 fhwLOl"
    >
      <div
        class="styles__Polygon-sc-1pwogyu-1 bFqjCw"
      >
        <div
          class="styles__Side-sc-1pwogyu-2 styles__Front-sc-1pwogyu-8 fiFRhu bmLGPK"
        >
          <div
            class="styles__Tracer-sc-18vfsm6-0 geFGiM"
            direction="left"
          />
        </div>
        <div
          class="styles__Side-sc-1pwogyu-2 styles__Back-sc-1pwogyu-3 fiFRhu hPnqZU"
        >
          <div
            class="styles__Tracer-sc-18vfsm6-0 eDtdvo"
          />
        </div>
        <div
          class="styles__Side-sc-1pwogyu-2 styles__Top-sc-1pwogyu-6 fiFRhu vbVpk"
        />
        <div
          class="styles__Side-sc-1pwogyu-2 styles__Bottom-sc-1pwogyu-7 fiFRhu cjImYD"
        >
          <div
            class="styles__Tracer-sc-18vfsm6-0 eDtdvo"
          />
        </div>
        <div
          class="styles__Side-sc-1pwogyu-2 styles__Left-sc-1pwogyu-5 fiFRhu enKYWw"
        >
          <div
            class="styles__Tracer-sc-18vfsm6-0 geFGiM"
            direction="left"
          />
        </div>
        <div
          class="styles__Side-sc-1pwogyu-2 styles__Right-sc-1pwogyu-4 fiFRhu kZyHbZ"
        >
          <div
            class="styles__Tracer-sc-18vfsm6-0 eDtdvo"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
